### ISSUE TITLE
Force update of mapping

### ISSUE THREAD

**Issue by hanneskaeufler** [elastic/elasticsearch-php#186]
Hi again,

is it possible to force an update of an indexes mapping? I´m having trouble with flickering tests in some test suites which is a pain. Tests often look like this:

```php

function setUp() {
    $this->removeIndex(); // deletes an index say my.index
    $this->createIndex(); // creates the index 
    $this->createMapping(); // puts a mapping into my.index to define type myEntity with field foo.sort
}

function testStuff() {
    $this->createEntity(); // create something that is indexed in my.index
    
    $this->refresh(); // refresh my.index

    $this->doSomething();  // calls a search operation on my.index down the stack somewhere with sorting on foo.sort

    $this->assertTrue(true); // you get the point
}
```

This seemingly randomly fails with `Parse Failure [No mapping found for [foo.sort] in order to sort on]` although the mapping works correctly in dev and production.

Trying to inspect the mapping before `$this->doSomething()` (which randomly fails) leads me to believe the mapping takes some time to settle? Is there some way to force update the mapping through the php client?

Thanks

**Comment by polyfractal** [comment]
It's probably not the mapping itself, since the mapping update only returns once all nodes have acknowledged the update.  The parse error is a bit of a red-herring.

The likely cause is that your index wasn't fully initialized prior to updating the mapping.  Initialization can take a few ms and often fouls up integration tests.  Since the index wasn't fully initialized, your mapping update would then fail.  But by time the search is executed, the index finally finished initializing...but without a proper mapping, so you get the parse exception.

Try adding a `wait_for_status=green` call after your index creation.  This call will block until the index is fully initialized:

```php
$client->cluster()->health(['wait_for_status' => 'green']); 
```

**Comment by hanneskaeufler** [comment]
Hm, for a moment I thought it would fix the flickering but unfortunately it does not :/ When I query the mapping

```php
$client->indices()->getMapping(['index' => 'my.index'])
```
before `$this->doSomething`, I will sometimes get the mapping, other times I wont.

**Comment by polyfractal** [comment]
Hmm...that's very odd.  What version of Elasticsearch are you using?  How many nodes are in your cluster?  Maybe you have a chronic split brain, so that there two "hemispheres" to your cluster, each with different sets of mappings?

**Comment by hanneskaeufler** [comment]
Hi, running 1.4.2

```json
{
  "status" : 200,
  "name" : "xxx",
  "cluster_name" : "xxx",
  "version" : {
    "number" : "1.4.2",
    "build_hash" : "927caff6f05403e936c20bf4529f144f0c89fd8c",
    "build_timestamp" : "2014-12-16T14:11:12Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.2"
  },
  "tagline" : "You Know, for Search"
}
```

With 2 nodes in the cluster. How would I investigate such a split brain? This is the cluster diagnostics from the "Elasticsearch HQ" interface:

![bildschirmfoto 2015-02-04 um 09 12 18](https://cloud.githubusercontent.com/assets/68024/6037007/84e4b550-ac4e-11e4-934e-55bdb769708a.png)
![bildschirmfoto 2015-02-04 um 09 12 34](https://cloud.githubusercontent.com/assets/68024/6037009/88e92924-ac4e-11e4-9791-e2d4f84cde2b.png)

**Comment by polyfractal** [comment]
Sorry for the delay, been sick :(  The easiest way to check for a split brain is to ask each node for it's local cluster state:

```
curl -XGET 'http://hostABC:9200/_cluster/state?pretty&local=true' | grep master
curl -XGET 'http://hostXYZ:9200/_cluster/state?pretty&local=true' | grep master
```

If there is a split brain, both of those nodes will report a different master (e.g. themselves).  If the master is the same, you're fine and it's back to the drawing board.

**Comment by hanneskaeufler** [comment]
No worries! So it´s not a split brain. But I´m still bugged by the issue all the time :(

**Comment by hanneskaeufler** [comment]
Might this be related to https://github.com/elastic/elasticsearch/issues/8688?

**Comment by polyfractal** [comment]
I doubt it tbh, since that issue is more about parallel updates to the same mapping from different threads of execution.  E.g. two nodes get two different mappings at the same time, both accept it locally, then have problems merging it later.  

I'm assuming you're running these tests single-threaded?

I've been noodling over this, and I'm still not sure *why* you're seeing such a drastic effect, but I think we may be able to beat it by checking the mapping on each node.  It'll be super hacky, but might at least allow forward progress.  something like:

```php
function setUp() {
    $this->removeIndex(); // deletes an index say my.index
    $this->createIndex(); // creates the index 
    $this->createMapping(); // puts a mapping into my.index to define type myEntity with field foo.sort
    
    // Wait for shards to initialize
    $client->cluster()->health(['wait_for_status' => 'green']); 
    
    foreach ($node in $nodes) {
      $mapping = [];
      do {
        // Get the local mapping from each node
        $mapping = $client->indices()->getMapping(['index' => 'my.index', 'local' => true]);
      } while (isset($mapping['path']['to']['field']) === false );
    }
    // All mapping updates applied
}
```

Basically, apply the mapping, wait for green, then manually verify that both nodes have applied the mapping of interest.  If these loops never finish spinning, there is something else wrong...not just a transient hiccup.

Do you have any other nodes running locally that could be interacting with the cluster?  E.g anything unexpected in:

```
ps aux | grep elasticsearch
```

**Comment by polyfractal** [comment]
Closing due to inactivity, please feel free to re-open if you have any more info.
